File: genkit-tools/cli/src/mcp/flows.ts
/**
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { RuntimeManager } from '@genkit-ai/tools-common/manager';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp';
import z from 'zod';

export function defineFlowTools(server: McpServer, manager: RuntimeManager) {
  server.registerTool(
    'list_flows',
    {
      title: 'List Genkit Flows',
      description:
        'Use this to discover available Genkit flows or inspect the input schema of Genkit flows to know how to successfully call them.',
    },
    async () => {
      const actions = await manager.listActions();

      let flows = '';
      for (const key of Object.keys(actions)) {
        if (key.startsWith('/flow/')) {
          flows += ' - Flow name: ' + key.substring('/flow/'.length) + '\n';
          if (actions[key].description) {
            flows += '   Description: ' + actions[key].description + '\n';
          }
          flows +=
            '   Input schema: ' +
            JSON.stringify(actions[key].inputSchema, undefined, 2) +
            '\n\n';
        }
      }

      return { content: [{ type: 'text', text: flows }] };
    }
  );

  server.registerTool(
    'run_flow',
    {
      title: 'Run Flow',
      description: 'Runs the flow with the provided input',
      inputSchema: {
        flowName: z.string().describe('name of the flow'),
        input: z
          .string()
          .describe(
            'Flow input as JSON object encoded as string (it will be passed through `JSON.parse`). Must conform to the schema.'
          )
          .optional(),
      },
    },
    async ({ flowName, input }) => {
      try {
        const response = await manager.runAction({
          key: `/flow/${flowName}`,
          input: input !== undefined ? JSON.parse(input) : undefined,
        });
        return {
          content: [
            { type: 'text', text: JSON.stringify(response, undefined, 2) },
          ],
        };
      } catch (e) {
        return {
          content: [{ type: 'text', text: `Error: ${e}` }],
        };
      }
    }
  );
}

File: py/samples/menu/src/case_04/flows.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0


"""Flows for case 04."""

import json
import os

from menu_ai import ai
from menu_schemas import AnswerOutputSchema, MenuItemSchema, MenuQuestionInputSchema
from pydantic import BaseModel, Field

from genkit.blocks.document import Document

from .prompts import s04_rag_data_menu_prompt


class IndexMenuItemsOutputSchema(BaseModel):
    """Output schema for indexing items."""

    rows: int = Field(...)


@ai.flow(name='s04_index_menu_items')
async def s04_index_menu_items_flow(
    menu_items: list[MenuItemSchema],
) -> IndexMenuItemsOutputSchema:
    """Index menu items for retrieval.

    Args:
        menu_items: List of menu items to index.

    Returns:
        Number of items indexed.

    Example:
        >>> await s04_index_menu_items_flow([MenuItemSchema(title='Burger', price=10.0, description='Yum')])
        IndexMenuItemsOutputSchema(rows=1)
    """
    # If empty list provided (e.g., from Dev UI default), load from example file
    if not menu_items:
        example_file = os.path.join(os.path.dirname(__file__), 'example.indexMenuItems.json')
        with open(example_file) as f:
            menu_data = json.load(f)
        menu_items = [MenuItemSchema(**item) for item in menu_data]

    documents = [
        Document.from_text(f'{item.title} {item.price} \n {item.description}', metadata=item.model_dump())
        for item in menu_items
    ]

    await ai.index(
        indexer='menu-items',
        documents=documents,
    )
    return IndexMenuItemsOutputSchema(rows=len(menu_items))


@ai.flow(name='s04_rag_menu_question')
async def s04_rag_menu_question_flow(
    my_input: MenuQuestionInputSchema,
) -> AnswerOutputSchema:
    """Answer a question using RAG on menu items.

    Args:
        my_input: Input containing the question.

    Returns:
        The answer.

    Example:
        >>> await s04_rag_menu_question_flow(MenuQuestionInputSchema(question='Do you have burgers?'))
        AnswerOutputSchema(answer="Yes, we have...")
    """
    # Retrieve the 3 most relevant menu items for the question
    docs = await ai.retrieve(
        retriever='menu-items',
        query=my_input.question,
        options={'limit': 3},
    )

    menu_data = [MenuItemSchema(**doc.metadata) for doc in docs.documents if doc.metadata]

    # Generate the response
    response = await s04_rag_data_menu_prompt({
        'menuData': [item.model_dump() for item in menu_data],
        'question': my_input.question,
    })
    return AnswerOutputSchema(answer=response.text)
